### PR TITLE
Skip sibling pipeline aggregators reduction during non-final reduce

### PR DESCRIPTION
**Pull request by javanna**
Today a coordinating node forces a final reduction of sibling pipeline aggregators whenever reducing aggs, unless it is reducing aggs incrementally. This works well for incremental reduction of aggs, but breaks CCS when minimizing roundtrips as each cluster ends up reducing its own pipeline aggregators locally while that should only be done by the CCS coordinating node later. This causes issues as after their reduction,  pipeline aggs cannot be further reduced, which is what happens with CCS causing errors like "java.lang.UnsupportedOperationException: Not supported" being returned.

Each coordinating node should rather honour the reduce context flag that
indicates whether we are executing a final reduce or not. If not, it should leave the sibling pipeline aggregations alone.

Note that his bug affects only pipeline aggs that don't have a parent in
the aggs tree, while all the others work well.

Relates to #40059 but does not fix it yet, as the CCS coordinating node also needs to be adapted to recreate sibling pipeline aggregators from the request.

### CODE REVIEW

**Review by jimczi — APPROVED** [review]
The change looks good to me but I'd like @polyfractal or @colings86 to take a look since they are more familiar with pipeline aggs.

File: server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java [hunk 1]
```diff
@@ -89,6 +97,15 @@ public static InternalAggregations reduce(List<InternalAggregations> aggregation
...
         }
+
+        if (siblingPipelineAggregators != null) {
+            if (context.isFinalReduce()) {
+                for (SiblingPipelineAggregator pipelineAggregator : siblingPipelineAggregators) {
+                    InternalAggregation newAgg = pipelineAggregator.doReduce(new InternalAggregations(reducedAggregations), context);
+                    reducedAggregations.add(newAgg);
+                }
+            }
+        }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java]: Note that this is moved from SearchPhaseController (and fixed to honour the finalReduce flag), as I want to be able to later call this from SearchResponseMerger 7.0+ only change)

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 2]
```diff
@@ -315,8 +315,6 @@ public void onFailure(Exception e) {
             if (localIndices != null) {
                 ActionListener<SearchResponse> ccsListener = createCCSListener(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY,
                     false, countDown, skippedClusters, exceptions, searchResponseMerger, totalClusters, listener);
-                //here we provide the empty string a cluster alias, which means no prefix in index name,
-                //but the coord node will perform non final reduce as it's not null.
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: this is an outdated comment, unrelated to this change, I just stumbled upon it while working on it.

**Review by polyfractal — APPROVED** [review]
Also LGTM.  

For posterity, I was curious why sibling aggs weren't throwing this exception normally under incremental reduce today, since it lacked the check.  `reduceAggs()` is invoked by incremental reductions but nulls out the sibling aggs first (to keep them from running), or invoked directly by the SearchPhaseController when we are not incrementally reducing (in which case the siblings are valid).

I like the new arrangement, a little clearer since it's always relying on the reduce context instead of nullness of siblings.

**Pull request merged**